File: test/hardware/unit/test_l2_cache.py
import cocotb
from cocotb.triggers import RisingEdge, Timer
from cocotb.clock import Clock
from pathlib import Path
from test.driver import run_hardware_test
from test.env import get_rtl_dir


@cocotb.test()
async def test_l2_cache_basic(dut):
    """Test L2 cache with simple read operations."""
    
    clock = Clock(dut.clk, 10, units="ns")
    cocotb.start_soon(clock.start())
    
    # Reset
    dut.rst_n.value = 0
    dut.s_en.value = 0
    dut.s_we.value = 0
    dut.mem_ready.value = 0
    dut.mem_rdata.value = 0
    await Timer(20, units="ns")
    await RisingEdge(dut.clk)
    dut.rst_n.value = 1
    await RisingEdge(dut.clk)
    
    # Test 1: Read miss
    dut.s_addr.value = 0x1000
    dut.s_we.value = 0
    dut.s_en.value = 1
    dut.s_be.value = 0b1111
    await RisingEdge(dut.clk)
    
    # Should not be ready immediately (miss)
    assert dut.s_ready.value == 0, "Should not be ready on miss"
    assert dut.mem_req.value == 1, "Should request memory"
    
    # Simulate memory responses for cache line fill (4 words)
    for i in range(4):
        dut.mem_rdata.value = 0x10000000 + (i << 8)
        dut.mem_ready.value = 1
        await RisingEdge(dut.clk)
        dut.mem_ready.value = 0
    
    # Wait for update
    await RisingEdge(dut.clk)
    
    # Test 2: Read hit
    dut.s_addr.value = 0x1000
    dut.s_en.value = 1
    dut.s_we.value = 0
    await RisingEdge(dut.clk)
    
    assert dut.s_ready.value == 1, "Should be ready on hit"
    assert dut.s_rdata.value == 0x10000000, "Should return cached data"
    
    dut._log.info("L2 Cache test passed")


def test_l2_cache():
    rtl_dir = get_rtl_dir()
    run_hardware_test(
        module_name=Path(__file__).stem,
        toplevel="l2_cache",
        verilog_sources=[rtl_dir / "cache" / "l2_cache.v"]
    )

File: test/hardware/unit/test_l2_cache.cpp
#include "tb_base.h"
#include "Vl2_cache.h"
#include <string>

class L2CacheTestbench : public ClockedTestbench<Vl2_cache> {
public:
    L2CacheTestbench() : ClockedTestbench<Vl2_cache>(200, true, "l2_cache_trace.vcd") {
        // Initialize inputs
        dut->s_en = 0;
        dut->s_we = 0;
        dut->s_addr = 0;
        dut->s_wdata = 0;
        dut->s_be = 0;
        dut->mem_ready = 0;
        dut->mem_rdata = 0;
    }
    
    void set_clk(uint8_t value) override {
        dut->clk = value;
    }
    
    void reset() {
        dut->rst_n = 0;
        tick();
        dut->rst_n = 1;
        tick();
        TB_LOG("Reset complete");
    }
    
    void test_read_miss() {
        TB_LOG("Test: L2 read miss and refill");
        
        // Read request
        dut->s_addr = 0x1000;
        dut->s_we = 0;
        dut->s_en = 1;
        dut->s_be = 0b1111;
        tick();
        
        // Should not be ready (miss)
        TB_ASSERT_EQ(dut->s_ready, 0, "Should not be ready on miss");
        TB_ASSERT_EQ(dut->mem_req, 1, "Should request memory");
        
        // Simulate memory responses for cache line fill
        for (int i = 0; i < 4; i++) {
            dut->mem_rdata = 0x10000000 + (i << 8);
            dut->mem_ready = 1;
            tick();
            dut->mem_ready = 0;
        }
        
        // Wait for update
        tick();
        
        // Read hit now
        dut->s_addr = 0x1000;
        dut->s_en = 1;
        dut->s_we = 0;
        tick();
        
        TB_ASSERT_EQ(dut->s_ready, 1, "Should be ready on hit");
        TB_ASSERT_EQ(dut->s_rdata, 0x10000000, "Cached data");
        
        dut->s_en = 0;
        tick();
    }
};

int main(int argc, char** argv) {
    Verilated::commandArgs(argc, argv);
    
    try {
        L2CacheTestbench tb;
        
        tb.reset();
        tb.test_read_miss();
        
        TB_LOG("All L2 Cache tests PASSED!");
        return 0;
        
    } catch (const std::exception& e) {
        TB_ERROR(e.what());
        return 1;
    }
}
